### ISSUE TITLE
Increase maximum salt length limit in OpenCL PBKDF2 code

### ISSUE THREAD

**Issue by kholia** [magnumripper/JohnTheRipper#3035]
The `OpenBSD-SoftRAID` format requires a salt length of 126 bytes (fixed).

Our current limit is 115 bytes as mentioned in `opencl_pbkdf2_hmac_sha1.h` file.

@jfoug Thanks for increasing the salt length limit to 115 last time. I need you again ;)

This issue blocks https://github.com/magnumripper/JohnTheRipper/issues/3021.

**Comment by jfoug** [comment]
Should this go to 179 bytes (3 limbs), or 243 bytes (4 limbs)?  note, this makes the salt structure (which is fixed size) larger, and forces more data transfer to the GPU. 

Also, does only pbkdf2-hmac-sha1 need expanded?

**Comment by jfoug** [comment]
Would this be better to have multiple salt structures, and the format pick the right one, so that the format itself knows the minimal size of salt object?  This could be done with a set of #defines I think.

**Comment by jfoug** [comment]
Something like this in opencl_pbdf2_hmac_sha1.h:

```
#if defined(OPENCL_PBKDF2_HMAC_SHA1_1LIMB)
typedef struct {
	unsigned int  length;
	unsigned int  outlen;
	unsigned int  iterations;
	unsigned char salt[55];
} pbkdf2_salt;
#elif defined (OPENCL_PBKDF2_HMAC_SHA1_2LIMB)
typedef struct {
	unsigned int  length;
	unsigned int  outlen;
	unsigned int  iterations;
	unsigned char salt[115];
} pbkdf2_salt;
#elif defined (OPENCL_PBKDF2_HMAC_SHA1_3LIMB)
typedef struct {
	unsigned int  length;
	unsigned int  outlen;
	unsigned int  iterations;
	unsigned char salt[179];
} pbkdf2_salt;
#elif defined (OPENCL_PBKDF2_HMAC_SHA1_4LIMB)
typedef struct {
	unsigned int  length;
	unsigned int  outlen;
	unsigned int  iterations;
	unsigned char salt[243];
} pbkdf2_salt;
#else
#error   DEFINE SOMETHING
#endif
```

**Comment by kholia** [comment]
@jfoug This sounds great to me.

The `OpenBSD-SoftRAID` format uses 128 byte salt.

**Comment by jfoug** [comment]
Here is my plan

* use defines for 2 or 3 limb salts (so the structure is minimized on formats NOT needing multi limb salts).
* add 3 limb salt logic to the kernel
* adjust any format files required.

I do have a question however.  Might it be of benefit to simply do the long salt computation using CTX model in the CPU code, and then simply always do a simple 1 limb salt within the kernel ?

**Comment by kholia** [comment]
I am not sure. Let's try to keep the initial version simple, and free of hacks. We can optimize later, if required :-)

**Comment by jfoug** [comment]
Ok.  I will simply do a 179 limit on salt (add the 3rd limb).  Not overly hard, but do you have test cases ready to make sure I do not Fuk something up?

**Comment by kholia** [comment]
I have an OpenCL format which will require 128 byte salt. I can use this to quickly test your code.

**Comment by kholia** [comment]
I am attaching the OpenCL format here as I might hit the bed soon.

[opencl_openbsdsoftraid_fmt_plug.c.txt](https://github.com/magnumripper/JohnTheRipper/files/1584223/opencl_openbsdsoftraid_fmt_plug.c.txt)

**Comment by magnumripper** [comment]
Haven't looked at anything yet but we should have a macro passed to the kernel build eg. `SALT_MAX_LEN`. The struct and code should adopt to it (virtually any number of limbs). Problem solved for good.

**Comment by jfoug** [comment]
Ok, I have this working, but not for the softraid format.  I expanded pbkdf2-hmac-sha1-opencl to handle 179 byte salts, and added a lot of tests to the common file.  I also split off the salt size from the 'common' 32 bit variant, since all of them were 2 limb.

Within the kernel, I replaced the code with generic 'works for any length' code.  I think we should put this code into all kernels, and from there on, salt length is only an issue on the CPU side.  The kernel is happy to do any salt.

Running a few more tests now, and I will get it checked in.   I am GLAD I made the extra tests (boundary lengths += 2).  It caught a bug that might have gotten through.  But I will make sample input file of salts of ALL lengths from 1 to 179 (IS ZERO length salt valid??), and test that the code finds ALL of them, before submitting.

**Comment by jfoug** [comment]
PR #3052 created.

**Comment by jfoug** [comment]
Done.

**Comment by jfoug** [comment]
Guess I will leave this open until I get the other ones done.

**Comment by jfoug** [comment]
I have all done, except sha256. I have that code done, but I do not think it is working.  I am letting the build bot go right now to see.  I will keep looking, but I might have to have someone else spot my obvious error in code.

**Comment by jfoug** [comment]
Can someone please look at e105056 and tell me what is wrong?

**Comment by jfoug** [comment]
Found problem.  there was a salt structure in the kernel .cl code, hard coded for 115 bytes.

I also changed the hmac loader to use the ctx model.  We should regression test to make sure I did not slow things down any.

**Comment by jfoug** [comment]
pbkdf2-sha256 opencl kinda sux.  I had to change several formats, each had structures which had to 'match'.  I converted that into a single header include for all CPU formats. For some formats, they extend the salt_t type.  So I put a salt_t into the formats structure, naming the formats structure salt_t2.  I also removed the 1 char for length, switching it to uint32. I also put the char* at the tail end of the structure.  Now we might be able to have slight changes to the structure (size of salt), since the other parts would still properly line up, AND we pass in the length, so no part of the salt past what the cpu part allocated will be accessed.  Right now, all the cpu parts, and the kernel all have a 179 byte salt buffer.  But I do not think it has to stay that way, and the kernel could possibly use a bigger buffer.
